server/android_notification_server.go

```diff
@@ -240,6 +234,73 @@ func (me *AndroidNotificationServer) SendNotification(msg *PushNotification) Pus
 	return NewOkPushResponse()
 }
 
+func (me *AndroidNotificationServer) SendNotificationWithRetry(fcmMsg *messaging.Message) error {
+	var err error
+	waitTime := time.Second
+
+	logger := me.logger.With(mlog.String("did", fcmMsg.Token))
+
+	// Keep a general context to make sure the whole retry
+	// doesn't take longer than the timeout.
+	generalContext, cancelGeneralContext := context.WithTimeout(context.Background(), me.sendTimeout)
+	defer cancelGeneralContext()
+
+	for retries := 0; retries < MAX_RETRIES; retries++ {
+		start := time.Now()
+
+		retryContext, cancelRetryContext := context.WithTimeout(generalContext, me.retryTimeout)
+		defer cancelRetryContext()
+		_, err = me.client.Send(retryContext, fcmMsg)
+		if me.metrics != nil {
+			me.metrics.observerNotificationResponse(PushNotifyAndroid, time.Since(start).Seconds())
+		}
+
+		if err == nil || !isRetryable(err) {
+			break
+		}
+
+		logger.Error(
+			"Failed to send android push",
+			mlog.Int("retry", retries),
+			mlog.Err(err),
+		)
+
+		if retries == MAX_RETRIES-1 {
+			logger.Error("Max retries reached")
+			break
+		}
+
+		select {
+		case <-generalContext.Done():
+		case <-time.After(waitTime):
+		}
+
+		if generalContext.Err() != nil {
+			logger.Info(
+				"Not retrying because context error",
+				mlog.Int("retry", retries),
+				mlog.Err(generalContext.Err()),
+			)
+			err = generalContext.Err()
+			break
+		}
```

Comment:
nit: Might we inline this within the `case <-generalContext.Done():`? When I first read it, I was thinking about the timer expiring vs. `generalContext` being done. `Err` just returns `nil` if it's not, so this code isn't wrong, but a more idiomatic pattern seems to be:

```go
select {
    case <-ctx.Done():
        return ctx.Err()
    case ...:
}
```

Reply:
I added the change, but I am 0/5 about the result. There is something that feels off about returning in a switch inside a loop. But maybe is just that it's been a long time since I do proper GO :P